normify.go

```diff
@@ -0,0 +1,299 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package wrp
+
+import (
+	"errors"
+	"strconv"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+var (
+	ErrInvalidMessageType = errors.New("invalid message type")
+	ErrInvalidPartnerID   = errors.New("invalid partner ID")
+	ErrInvalidSource      = errors.New("invalid source locator")
+	ErrInvalidDest        = errors.New("invalid destination locator")
+	ErrInvalidString      = errors.New("invalid UTF-8 string")
+)
+
+// Normify applies a series of normalizing options to a WRP message.
+type Normify struct {
+	opts []NormifyOption
+}
+
+// NormifyOption is a functional option for normalizing a WRP message.
+type NormifyOption interface {
+	// normify applies the option to the given message.
+	normify(*Message) error
+}
+
+// optionFunc is an adapter to allow the use of ordinary functions as
+// normalizing options.
+type optionFunc func(*Message) error
+
+var _ NormifyOption = optionFunc(nil)
+
+func (f optionFunc) normify(m *Message) error {
+	return f(m)
+}
+
+// New creates a new Correctifier with the given options.
+func New(opts ...NormifyOption) *Normify {
+	return &Normify{
+		opts: opts,
+	}
+}
+
+// Process applies the normalizing and validating options to the message.  It
+// returns an error if any of the options fail.
+func (n *Normify) Process(m *Message) error {
```

Comment:
style suggestion if we wanted to keep things arbitrarily consistent
```suggestion
func (n *Normify) Normify(m *Message) error {
```

or leave that along and change `func (f optionFunc) normify(m *Message) error ` to
```
func (f optionFunc) process(m *Message) error {
	return f(m)
}
```

Reply:
I'm fine with leaving it as is, but I wanted to point it out in case that was the intention

Reply:
Good question.  I was worried this might feel like a stuttering API if I called it `Normify`.  I'm fine either way.  @johnabass what do you think?

Reply:
The general convention with a single-method interface is essentially:

```
type Xer interface {
    X()
}
```

So, in this case:

```
type Normifier interface {
    Normify()
}
```

The stuttering principal doesn't apply because you don't refer to the method in the same line as the type in normal flow:

```
var n wrp.Normifier = wrp.NewNormifier(...)
n.Normify() // different line from above
```

Reply:
On a different note, where validation is taking place in this package should we consider reusing the wrpvalidation code where we can in order to keep things aligned?

Reply:
Yes, as much as possible we should use this library for validation.  There are some server-specific rules, like what scytale does.  But for the most part, all our services should apply the same validation.

Reply:
@denopink I think we should have `wrpvalidation` use this code when possible as `wrpvalidation` is quite involved regarding metrics.  I decided to punt on variants of these options that support metrics for now as I don't immediately need them nor do I have the cycles to refactor `wrpvalidation` to use this yet.  My current thinking on the pattern for how to do them is like this:

```golang
// EnsureTransactionUUID ensures that the message has a transaction UUID.  If
// the message does not have a transaction UUID, a new one is generated and
// added to the message.
func EnsureTransactionUUID(onadded ...func()) NormifierOption {
	return optionFunc(func(m *Message) error {
		if m.TransactionUUID == "" {
			id, err := uuid.NewRandom()
			if err != nil {
				return err
			}

			m.TransactionUUID = id.String()
			for _, fn := range onadded {
				if fn != nil {
					fn()
				}
			}
		}
		return nil
	})
}
```